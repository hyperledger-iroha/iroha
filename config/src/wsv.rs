//! Module for `WorldStateView`-related configuration and structs.
use default::*;
use iroha_config_base::derive::{Documented, Proxy};
use iroha_data_model::{prelude::*, transaction::TransactionLimits};
use serde::{Deserialize, Serialize};

use crate::wasm;

/// Module with a set of default values.
pub mod default {
    use super::*;

    /// Default limits for metadata
    pub const DEFAULT_METADATA_LIMITS: MetadataLimits =
        MetadataLimits::new(2_u32.pow(20), 2_u32.pow(12));
    /// Default limits for ident length
    pub const DEFAULT_IDENT_LENGTH_LIMITS: LengthLimits = LengthLimits::new(1, 2_u32.pow(7));
    /// Default maximum number of instructions and expressions per transaction
    pub const DEFAULT_MAX_INSTRUCTION_NUMBER: u64 = 2_u64.pow(12);
    /// Default maximum number of instructions and expressions per transaction
    pub const DEFAULT_MAX_WASM_SIZE_BYTES: u64 = 2_u64.pow(22); // 4 MiB

    /// Default transaction limits
    pub const DEFAULT_TRANSACTION_LIMITS: TransactionLimits =
        TransactionLimits::new(DEFAULT_MAX_INSTRUCTION_NUMBER, DEFAULT_MAX_WASM_SIZE_BYTES);
}

/// `WorldStateView` configuration.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize, Proxy, Documented)]
#[config(env_prefix = "WSV_")]
#[serde(rename_all = "UPPERCASE")]
pub struct Configuration {
    /// [`MetadataLimits`] for every asset with store.
    pub asset_metadata_limits: MetadataLimits,
    /// [`MetadataLimits`] of any asset definition metadata.
    pub asset_definition_metadata_limits: MetadataLimits,
    /// [`MetadataLimits`] of any account metadata.
    pub account_metadata_limits: MetadataLimits,
    /// [`MetadataLimits`] of any domain metadata.
    pub domain_metadata_limits: MetadataLimits,
    /// [`LengthLimits`] for the number of chars in identifiers that can be stored in the WSV.
    pub ident_length_limits: LengthLimits,
    /// Limits that all transactions need to obey, in terms of size
    /// of WASM blob and number of instructions.
    pub transaction_limits: TransactionLimits,
    /// WASM runtime configuration
    #[config(inner)]
    pub wasm_runtime_config: wasm::Configuration,
}

impl Default for ConfigurationProxy {
    fn default() -> Self {
        Self {
            asset_metadata_limits: Some(DEFAULT_METADATA_LIMITS),
            asset_definition_metadata_limits: Some(DEFAULT_METADATA_LIMITS),
            account_metadata_limits: Some(DEFAULT_METADATA_LIMITS),
            domain_metadata_limits: Some(DEFAULT_METADATA_LIMITS),
            ident_length_limits: Some(DEFAULT_IDENT_LENGTH_LIMITS),
            transaction_limits: Some(DEFAULT_TRANSACTION_LIMITS),
            wasm_runtime_config: Some(wasm::ConfigurationProxy::default()),
        }
    }
}

#[cfg(test)]
pub mod tests {
    use proptest::prelude::*;

    use super::*;

    prop_compose! {
        pub fn arb_proxy()
            (
                asset_metadata_limits in prop::option::of(Just(DEFAULT_METADATA_LIMITS)),
                asset_definition_metadata_limits in prop::option::of(Just(DEFAULT_METADATA_LIMITS)),
                account_metadata_limits in prop::option::of(Just(DEFAULT_METADATA_LIMITS)),
                domain_metadata_limits in prop::option::of(Just(DEFAULT_METADATA_LIMITS)),
                ident_length_limits in prop::option::of(Just(DEFAULT_IDENT_LENGTH_LIMITS)),
                transaction_limits in prop::option::of(Just(DEFAULT_TRANSACTION_LIMITS)),
                wasm_runtime_config in prop::option::of(Just(wasm::ConfigurationProxy::default())),
            )
            -> ConfigurationProxy {
            ConfigurationProxy { asset_metadata_limits, asset_definition_metadata_limits, account_metadata_limits, domain_metadata_limits, ident_length_limits, transaction_limits, wasm_runtime_config }
        }
    }
}
