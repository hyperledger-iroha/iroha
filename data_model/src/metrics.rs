//! [`Metrics`] and [`Status`]-related logic and functions.
use std::{sync::Arc, time::Duration};

use prometheus::{
    core::{AtomicU64, GenericGauge},
    Encoder, IntCounter, IntCounterVec, Opts, Registry,
};
use serde::{Deserialize, Serialize};

/// Thin wrapper around duration that `impl`s [`Default`]
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct Uptime(Duration);

impl Default for Uptime {
    fn default() -> Self {
        Self(Duration::from_millis(0))
    }
}

/// Response body for GET status request
#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
pub struct Status {
    /// Number of connected peers, except for the reporting peer itself
    pub peers: u64,
    /// Number of committed blocks
    pub blocks: u64,
    /// Number of transactions committed in the last block
    pub txs: u64,
    /// Uptime since genesis block creation
    pub uptime: Uptime,
}

impl From<&Arc<Metrics>> for Status {
    fn from(val: &Arc<Metrics>) -> Self {
        Self {
            peers: val.connected_peers.get(),
            blocks: val.block_height.get(),
            txs: val.txs.with_label_values(&["total"]).get(),
            uptime: Uptime(Duration::from_millis(val.uptime_since_genesis_ms.get())),
        }
    }
}

/// A strict superset of [`Status`].
#[derive(Debug)]
pub struct Metrics {
    /// Transactions in the last committed block
    pub txs: IntCounterVec,
    /// Current block height
    pub block_height: IntCounter,
    /// Total number of currently connected peers
    pub connected_peers: GenericGauge<AtomicU64>,
    /// Uptime of the network, starting from commit of the genesis block
    pub uptime_since_genesis_ms: GenericGauge<AtomicU64>,
    /// Number of domains.
    pub domains: GenericGauge<AtomicU64>,
    /// Number of users with non-zero assets.
    pub users: GenericGauge<AtomicU64>,
    /// Queries handled by this peer
    pub queries: IntCounterVec,
    // Internal use only.
    registry: Registry,
}

impl Default for Metrics {
    // The constructors either always fail, or never.
    #[allow(clippy::expect_used)]
    fn default() -> Self {
        let txs = IntCounterVec::new(Opts::new("txs", "Transactions committed"), &["type"])
            .expect("Infallible");
        let queries = IntCounterVec::new(
            Opts::new("queries", "Queries handled by this peer"),
            &["type", "success_status"],
        )
        .expect("Infallible");
        let block_height =
            IntCounter::new("block_height", "Current block height").expect("Infallible");
        let connected_peers = GenericGauge::new(
            "connected_peers",
            "Total number of currently connected peers",
        )
        .expect("Infallible");
        let uptime_since_genesis_ms = GenericGauge::new(
            "uptime_since_genesis_ms",
            "Network uptime, from creation of the genesis block",
        )
        .expect("Infallible");
        let domains = GenericGauge::new("domains", "Total number of domains").expect("Infallible");
        let users = GenericGauge::new("users", "Total number of users").expect("Infallible");

        let registry = Registry::new();

        macro_rules! register {
			($metric:expr)=> {
				registry.register(Box::new($metric.clone())).expect("Infallible");
			};
			($metric:expr,$($metrics:expr),+)=>{
				register!($metric);
				register!($($metrics),+);
			}
		}

        register!(
            txs,
            block_height,
            connected_peers,
            uptime_since_genesis_ms,
            domains,
            users,
            queries
        );

        Self {
            txs,
            block_height,
            connected_peers,
            uptime_since_genesis_ms,
            registry,
            domains,
            users,
            queries,
        }
    }
}

impl Metrics {
    /// Convert the current [`Metrics`] into a Prometheus-readable format.
    ///
    /// # Errors
    /// - If [`Encoder`] fails to encode the data
    /// - If the buffer produced by [`Encoder`] causes [`String::from_utf8`] to fail.
    pub fn try_to_string(&self) -> eyre::Result<String> {
        let mut buffer = vec![];
        let encoder = prometheus::TextEncoder::new();
        let metric_families = self.registry.gather();
        Encoder::encode(&encoder, &metric_families, &mut buffer)?;
        Ok(String::from_utf8(buffer)?)
    }
}
