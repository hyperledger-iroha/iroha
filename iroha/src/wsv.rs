//! This module provides `WorldStateView` - in-memory representations of the current blockchain
//! state.

use crate::prelude::*;

/// Current state of the blockchain alligned with `Iroha` module.
#[derive(Debug, Clone)]
pub struct WorldStateView {
    peer: Peer,
}

impl WorldStateView {
    /// Default `WorldStateView` constructor.
    pub fn new(peer: Peer) -> Self {
        WorldStateView { peer }
    }

    /// Put `ValidBlock` of information with changes in form of **Iroha Special Instructions**
    /// into the world.
    pub async fn put(&mut self, block: &CommittedBlock) {
        for transaction in &block.transactions {
            if let Err(e) = &transaction.proceed(self) {
                eprintln!("Failed to procced transaction on WSV: {}", e);
            }
        }
    }

    /// Get `Peer` with an ability to modify it.
    pub fn peer(&mut self) -> &mut Peer {
        &mut self.peer
    }

    /// Add new `Domain` entity.
    pub fn add_domain(&mut self, domain: Domain) {
        self.peer.domains.insert(domain.name.clone(), domain);
    }

    /// Get `Domain` without an ability to modify it.
    pub fn read_domain(&self, name: &str) -> Option<&Domain> {
        self.peer.domains.get(name)
    }

    /// Get `Domain` with an ability to modify it.
    pub fn domain(&mut self, name: &str) -> Option<&mut Domain> {
        self.peer.domains.get_mut(name)
    }

    /// Get `Account` without an ability to modify it.
    pub fn read_account(&self, id: &<Account as Identifiable>::Id) -> Option<&Account> {
        self.read_domain(&id.domain_name)?.accounts.get(id)
    }

    /// Get `Account` with an ability to modify it.
    pub fn account(&mut self, id: &<Account as Identifiable>::Id) -> Option<&mut Account> {
        self.domain(&id.domain_name)?.accounts.get_mut(id)
    }

    /// Get `Asset` without an ability to modify it.
    pub fn read_asset(&mut self, id: &<Asset as Identifiable>::Id) -> Option<&mut Asset> {
        self.account(&id.account_id)?.assets.get_mut(id)
    }

    /// Get `Asset` with an ability to modify it.
    pub fn asset(&mut self, id: &<Asset as Identifiable>::Id) -> Option<&mut Asset> {
        self.account(&id.account_id)?.assets.get_mut(id)
    }

    /// Add new `Asset` entity.
    pub fn add_asset(&mut self, asset: Asset) {
        self.account(&asset.id.account_id)
            .expect("Failed to find an account.")
            .assets
            .insert(asset.id.clone(), asset);
    }

    /// Get `AssetDefinition` without an ability to modify it.
    pub fn read_asset_definition(
        &self,
        id: &<AssetDefinition as Identifiable>::Id,
    ) -> Option<&AssetDefinition> {
        self.read_domain(&id.domain_name)?.asset_definitions.get(id)
    }

    /// Get `AssetDefinition` with an ability to modify it.
    pub fn asset_definition(
        &mut self,
        id: &<AssetDefinition as Identifiable>::Id,
    ) -> Option<&mut AssetDefinition> {
        self.domain(&id.domain_name)?.asset_definitions.get_mut(id)
    }
}
