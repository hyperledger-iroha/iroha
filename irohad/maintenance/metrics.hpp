/**
 * Copyright Soramitsu Co., Ltd. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#ifndef IROHA_MAINTENANCE_METRICS_HPP
#define IROHA_MAINTENANCE_METRICS_HPP

#include <string>
#include <memory>
#include <prometheus/registry.h>

std::shared_ptr<prometheus::Registry>
    maintenance_metrics_init(std::string const& listen_addr);

#endif //IROHA_MAINTENANCE_METRICS_HPP
