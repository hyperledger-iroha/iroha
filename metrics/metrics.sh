#!/usr/bin/env bash
set -euo pipefail

cd $(dirname $(realpath ${BASH_SOURCE[0]}))
prometheus --config.file=prometheus.yml
