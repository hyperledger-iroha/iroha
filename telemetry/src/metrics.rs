//! [`Metrics`] and [`Status`]-related logic and functions.

use std::{
    ops::Deref,
    time::{Duration, SystemTime},
};

use prometheus::{
    core::{AtomicU64, GenericGauge, GenericGaugeVec},
    Encoder, Histogram, HistogramOpts, HistogramVec, IntCounter, IntCounterVec, Opts, Registry,
};
use serde::{Deserialize, Serialize};

/// Thin wrapper around duration that `impl`s [`Default`]
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct Uptime(pub Duration);

impl Default for Uptime {
    fn default() -> Self {
        Self(Duration::from_millis(0))
    }
}

/// Response body for GET status request
#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
pub struct Status {
    /// Number of connected peers, except for the reporting peer itself
    pub peers: u64,
    /// Number of committed blocks
    pub blocks: u64,
    /// Number of accepted transactions
    pub txs_accepted: u64,
    /// Number of rejected transactions
    pub txs_rejected: u64,
    /// Uptime since genesis block creation
    pub uptime: Uptime,
    /// Number of view changes in the current round
    pub view_changes: u64,
    /// Number of the transactions in the queue
    pub queue_size: u64,
}

impl<T: Deref<Target = Metrics>> From<&T> for Status {
    fn from(value: &T) -> Self {
        let val: &Metrics = value;
        Self {
            peers: val.connected_peers.get(),
            blocks: val.block_height.get(),
            txs_accepted: val.txs.with_label_values(&["accepted"]).get(),
            txs_rejected: val.txs.with_label_values(&["rejected"]).get(),
            uptime: Uptime(Duration::from_millis(val.uptime_since_genesis_ms.get())),
            view_changes: val.view_changes.get(),
            queue_size: val.queue_size.get(),
        }
    }
}

/// A strict superset of [`Status`].
#[derive(Debug, Clone)]
pub struct Metrics {
    /// Total number of transactions
    pub txs: IntCounterVec,
    /// Current block height
    pub block_height: IntCounter,
    /// Total number of currently connected peers
    pub connected_peers: GenericGauge<AtomicU64>,
    /// Uptime of the network, starting from commit of the genesis block
    pub uptime_since_genesis_ms: GenericGauge<AtomicU64>,
    /// Number of domains.
    pub domains: GenericGauge<AtomicU64>,
    /// Total number of users per domain
    pub accounts: GenericGaugeVec<AtomicU64>,
    /// Transaction amounts.
    pub tx_amounts: Histogram,
    /// Queries handled by this peer
    pub isi: IntCounterVec,
    /// Query handle time Histogram
    pub isi_times: HistogramVec,
    /// Number of view changes in the current round
    pub view_changes: GenericGauge<AtomicU64>,
    /// Number of transactions in the queue
    pub queue_size: GenericGauge<AtomicU64>,
    /// Number of sumeragi dropped messages
    pub dropped_messages: IntCounter,
    /// Internal use only. Needed for generating the response.
    registry: Registry,
}

impl Default for Metrics {
    fn default() -> Self {
        let txs = IntCounterVec::new(Opts::new("txs", "Transactions committed"), &["type"])
            .expect("Infallible");
        let isi = IntCounterVec::new(
            Opts::new("isi", "Iroha special instructions handled by this peer"),
            &["type", "success_status"],
        )
        .expect("Infallible");
        let isi_times = HistogramVec::new(
            HistogramOpts::new("isi_times", "Time to handle isi in this peer"),
            &["type"],
        )
        .expect("Infallible");
        let tx_amounts = Histogram::with_opts(HistogramOpts::new(
            "tx_amount",
            "average amount involved in a transaction on this peer",
        ))
        .expect("Infallible");
        let block_height =
            IntCounter::new("block_height", "Current block height").expect("Infallible");
        let connected_peers = GenericGauge::new(
            "connected_peers",
            "Total number of currently connected peers",
        )
        .expect("Infallible");
        let uptime_since_genesis_ms = GenericGauge::new(
            "uptime_since_genesis_ms",
            "Network up-time, from creation of the genesis block",
        )
        .expect("Infallible");
        let domains = GenericGauge::new("domains", "Total number of domains").expect("Infallible");
        let accounts = GenericGaugeVec::new(
            Opts::new("accounts", "User accounts registered at this time"),
            &["domain"],
        )
        .expect("Infallible");
        let view_changes = GenericGauge::new(
            "view_changes",
            "Number of view changes in the current round",
        )
        .expect("Infallible");
        let queue_size = GenericGauge::new("queue_size", "Number of the transactions in the queue")
            .expect("Infallible");
        let dropped_messages =
            IntCounter::new("dropped_messages", "Sumeragi dropped messages").expect("Infallible");
        let registry = Registry::new();

        macro_rules! register {
            ($metric:expr)=> {
                registry.register(Box::new($metric.clone())).expect("Infallible");
            };
            ($metric:expr,$($metrics:expr),+)=>{
                register!($metric);
                register!($($metrics),+);
            }
        }

        register!(
            txs,
            tx_amounts,
            block_height,
            connected_peers,
            uptime_since_genesis_ms,
            domains,
            accounts,
            isi,
            isi_times,
            view_changes,
            queue_size,
            dropped_messages
        );

        Self {
            txs,
            block_height,
            connected_peers,
            uptime_since_genesis_ms,
            domains,
            accounts,
            tx_amounts,
            isi,
            isi_times,
            view_changes,
            queue_size,
            dropped_messages,
            registry,
        }
    }
}

impl Metrics {
    /// Convert the current [`Metrics`] into a Prometheus-readable format.
    ///
    /// # Errors
    /// - If [`Encoder`] fails to encode the data
    /// - If the buffer produced by [`Encoder`] causes [`String::from_utf8`] to fail.
    pub fn try_to_string(&self) -> eyre::Result<String> {
        let mut buffer = Vec::new();
        let encoder = prometheus::TextEncoder::new();
        let metric_families = self.registry.gather();
        Encoder::encode(&encoder, &metric_families, &mut buffer)?;
        Ok(String::from_utf8(buffer)?)
    }

    /// Get time elapsed since Unix epoch.
    ///
    /// # Panics
    /// Never
    #[allow(clippy::unused_self)]
    pub fn current_time(&self) -> Duration {
        SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .expect("Failed to get the current system time")
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn metrics_lifecycle() {
        let metrics = Metrics::default();
        println!(
            "{:?}",
            metrics
                .try_to_string()
                .expect("Should not fail for default")
        );
        println!("{:?}", Status::from(&Box::new(metrics)));
        println!("{:?}", Status::default());
    }
}
