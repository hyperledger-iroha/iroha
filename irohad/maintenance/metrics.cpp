/**
 * Copyright Soramitsu Co., Ltd. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "maintenance/metrics.hpp"

#include <prometheus/counter.h>
#include <prometheus/exposer.h>
#include <prometheus/registry.h>

#include <array>
#include <chrono>
#include <cstdlib>
#include <memory>
#include <string>
#include <thread>
#include <future>
#include <regex>

std::shared_ptr<prometheus::Registry> maintenance_metrics_init(std::string const& listen_addr)
{
  using namespace prometheus;

  static const std::regex full_matcher("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]):[0-9]+$");
  static const std::regex port_matcher("^:?([0-9]{1,5})$");
  std::string listen_addr_port;
  if(std::regex_match(listen_addr,full_matcher)) {
    listen_addr_port = listen_addr;
  } else if(std::regex_match(listen_addr,port_matcher)) {
    listen_addr_port = "127.0.0.1";
    if (listen_addr[0] != ':')
      listen_addr_port += ":";
    listen_addr_port += listen_addr;
  } else {
    return nullptr;
  }

  // create a metrics registry
  // @note it's the users responsibility to keep the object alive
  auto registry = std::make_shared<Registry>();

  // Just for example
  std::thread([registry,listen_addr_port](){
      // create an http server running on addr:port
      Exposer exposer{listen_addr_port};

      // ask the exposer to scrape the registry on incoming HTTP requests
      exposer.RegisterCollectable(registry);

    // add a new counter family to the registry (families combine values with the
    // same name, but distinct label dimensions)
    //
    // @note please follow the metric-naming best-practices:
    // https://prometheus.io/docs/practices/naming/
    auto& packet_counter = BuildCounter()
        .Name("observed_packets_total")
        .Help("Number of observed packets")
        .Register(*registry);

    // add and remember dimensional data, incrementing those is very cheap
    auto& tcp_rx_counter =
        packet_counter.Add({{"protocol", "tcp"}, {"direction", "rx"}});
    auto& tcp_tx_counter =
        packet_counter.Add({{"protocol", "tcp"}, {"direction", "tx"}});
    auto& udp_rx_counter =
        packet_counter.Add({{"protocol", "udp"}, {"direction", "rx"}});
    auto& udp_tx_counter =
        packet_counter.Add({{"protocol", "udp"}, {"direction", "tx"}});

    // add a counter whose dimensional data is not known at compile time
    // nevertheless dimensional values should only occur in low cardinality:
    // https://prometheus.io/docs/practices/naming/#labels
    auto& http_requests_counter = BuildCounter()
        .Name("http_requests_total")
        .Help("Number of HTTP requests")
        .Register(*registry);

    for (;;) {
      std::this_thread::sleep_for(std::chrono::seconds(1));
      const auto random_value = std::rand();

      if (random_value & 1) tcp_rx_counter.Increment();
      if (random_value & 2) tcp_tx_counter.Increment();
      if (random_value & 4) udp_rx_counter.Increment();
      if (random_value & 8) udp_tx_counter.Increment();

      const std::array<std::string, 4> methods = {"GET", "PUT", "POST", "HEAD"};
      auto method = methods.at(random_value % methods.size());
      // dynamically calling Family<T>.Add() works but is slow and should be avoided
      http_requests_counter.Add({{"method", method}}).Increment();
    }
  }).detach();

  return {registry};
}
