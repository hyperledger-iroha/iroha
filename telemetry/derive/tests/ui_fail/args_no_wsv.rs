use iroha_core::wsv::{World, WorldStateView};
use iroha_telemetry_derive::metrics;

#[metrics(+"test_query", "another_test_query_without_timing")]
fn execute(_wsv: &World) -> Result<(), ()> {
    Ok(())
}

fn main() {
    let _world = WorldStateView::<World>::default();
}
