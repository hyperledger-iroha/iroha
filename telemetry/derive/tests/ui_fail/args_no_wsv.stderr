error: At least one argument must be a `WorldStateView`.
 --> tests/ui_fail/args_no_wsv.rs:4:12
  |
4 | fn execute(_wsv: &World) -> Result<(), ()> {
  |            ^^^^^^^^^^^^
